### ISSUE TITLE
GPT-NeoX: hardcoded float32 softmax silently caps float64 inference at float32 resolution

### ISSUE THREAD

**Issue by clashstar** [huggingface/transformers#48459]
### System Info

- `transformers` version: 5.14.1
- Platform: Windows-11-10.0.26200-SP0
- Python version: 3.12.10
- Huggingface_hub version: 1.25.1
- Safetensors version: 0.8.0
- Accelerate version: 1.14.0
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.13.0+cpu (NA)
- Using distributed or parallel set-up in script?: No

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

The eager attention path for GPT-NeoX hardcodes the attention softmax dtype:

```python
attn_weights = nn.functional.softmax(
    attn_weights, dim=-1, dtype=torch.float32
).to(query.dtype)
```

The issue is specifically the float64 case.

For float16/bfloat16, forcing the softmax to float32 is a reasonable numerical-stability upcast. For float32 it is effectively a no-op. However, when a GPT-NeoX model is explicitly converted to float64, this line instead silently downcasts every attention map to float32 before casting the result back to float64.

As a result, the model can expose float64 parameters, activations, and outputs while part of the forward pass is still capped at float32 resolution.

### Minimal reproduction

Using Pythia-410M-deduped with eager attention:

1. Convert all model parameters and floating-point inputs to float64.
2. Run a forward pass and record the final logits.
3. Add a perturbation of magnitude `1e-9` to a mid-stream residual activation through a forward hook.
4. Run the forward pass again.
5. Compare the resulting logits.

With the current GPT-NeoX eager-attention implementation, the response to sufficiently small perturbations is annihilated by the float32 softmax path.

Changing the softmax to:

```python
attn_weights = nn.functional.softmax(
    attn_weights, dim=-1, dtype=query.dtype
)
```

restores the expected small-perturbation response in float64.

An alternative that preserves the existing stability policy for lower-precision inference while avoiding a float64 downcast would be:

```python
dtype = torch.promote_types(query.dtype, torch.float32)

attn_weights = nn.functional.softmax(
    attn_weights, dim=-1, dtype=dtype
).to(query.dtype)
```

### Numerical evidence

This was discovered while measuring finite activation-patching interactions, which involve small four-term differences and are particularly sensitive to per-forward rounding.

For Pythia-410M in nominal float64:

* current implementation: interaction measurements develop a numerical floor at approximately `5e-5` logit-difference units;
* dtype-preserving implementation: interactions remain resolvable to approximately `1e-6`;
* the corrected path also recovers the expected small-displacement quadratic scaling that is obscured by the current precision floor.

For float32 inference, we additionally tested 201 randomized draws and found bitwise-identical outputs between the original behavior and a minimum-float32/promoted-dtype implementation.

The relevant line remains present in the current `main` GPT-NeoX eager-attention implementation.

Full provenance, including the probe, original function, unified diff, and float32 identity tests, is available at:

https://doi.org/10.5281/zenodo.21736208

See `data/patch_provenance/`.

### Expected behavior

A model whose parameters and floating-point inputs are explicitly converted to float64 should not silently perform a numerically significant intermediate operation at float32 precision unless this limitation is documented.

The existing float32 softmax appears reasonable as a numerical-stability policy for float16/bfloat16 inference. The issue is that the same policy becomes a precision-reducing downcast when `query.dtype == torch.float64`.

A dtype-promoting implementation such as:

```python
dtype = torch.promote_types(query.dtype, torch.float32)
```

would preserve the current minimum-float32 stability behavior for lower-precision models while allowing genuine float64 computation when float64 is explicitly requested.

If maintainers consider the existing behavior intentional, documenting that GPT-NeoX eager attention is capped at float32 softmax precision during float64 inference would also resolve the ambiguity.

I am happy to open a PR if maintainers agree that the dtype-promoting behavior is preferable.

**Comment by Mahnoor-Zaffar** [comment]
Hey! Thanks for reporting this.

The core problem seems to stem from the line in the eager attention path where the softmax is called with a fixed `dtype` of `torch.float32`. This is located in the attention mechanism implementation, specifically in the function handling the attention weights. The relevant code is:

```python
attn_weights = nn.functional.softmax(
    attn_weights, dim=-1, dtype=torch.float32
).to(query.dtype)
```

This line does not account for the case when `query.dtype` is `torch.float64`, resulting in a silent downcast to `float32`.

To address this issue, the softmax function should be modified to use the same dtype as the query tensor, or a dtype promotion strategy could be implemented to ensure that the precision is preserved during the softmax operation.

If no one is currently working on this, I'd be happy to take a look and put together a fix.

**Comment by vasqu** [comment]
I'm honestly not sure whether we even want this. There is no real reason to support fp64 dtype imo

If we were to support we would need to rethink essentially every model again especially since mps is also very weird around fp64. I would close this as not planned tbh
